### PR TITLE
Validation cleanup

### PR DESCRIPTION
**Pull request by Shane32**

### CODE REVIEW

File: src/GraphQL/Validation/Rules/ArgumentsOfCorrectType.cs [hunk 1]
```diff
@@ -17,9 +17,7 @@ public class ArgumentsOfCorrectType : IValidationRule
 
         public Task<INodeVisitor> ValidateAsync(ValidationContext context)
         {
-            return new EnterLeaveListener(_ =>
-            {
-                _.Match<Argument>(argAst =>
+            return new MatchingNodeVisitor<Argument>(argAst =>
```
Comment (sungam3r) [inline src/GraphQL/Validation/Rules/ArgumentsOfCorrectType.cs:20]: Less memory allocations 👍

File: src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs [hunk 2]
```diff
@@ -723,26 +723,26 @@ private bool SameValue(Argument arg1, Argument arg2)
             }
         }
 
-        public class FieldDefPair
+        private class FieldDefPair
```
Comment (sungam3r) [inline src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs:726]:
```suggestion
        private sealed class FieldDefPair
```

File: src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs [hunk 3]
```diff
@@ -795,13 +795,13 @@ public class Message
             public List<ConflictReason> Msgs { get; set; }
         }
 
-        public class CachedField
+        private class CachedField
```
Comment (sungam3r) [inline src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs:798]:
```suggestion
        private sealed class CachedField
```

File: src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs [hunk 4]
```diff
@@ -795,13 +795,13 @@ public class Message
...
         {
             public Dictionary<string, List<FieldDefPair>> NodeAndDef { get; set; }
             public List<string> Names { get; set; }
         }
 
-        public class PairSet
+        private class PairSet
```
Comment (sungam3r) [inline src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs:804]:
```suggestion
        private sealed class PairSet
```

File: src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs [hunk 5]
```diff
@@ -848,7 +848,7 @@ private void PairSetAdd(string a, string b, bool areMutuallyExclusive)
             }
         }
 
-        public class ObjMap<T> : Dictionary<string, T>
+        private class ObjMap<T> : Dictionary<string, T>
```
Comment (sungam3r) [inline src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs:851]:
```suggestion
        private sealed class ObjMap<T> : Dictionary<string, T>
```

**Comment by sungam3r** [comment]
@Shane32 You gave me an interesting idea of how we can further improve the validation. I would like to continue on your branch.

**Comment by Shane32** [comment]
@sungam3r Can we merge this first? I'd like to first put in a PR for further elimination of memory allocations.  And then you can have a go at it.

File: src/GraphQL/Validation/Rules/ArgumentsOfCorrectType.cs (hunk 1 above)
Comment (Shane32) [reply src/GraphQL/Validation/Rules/ArgumentsOfCorrectType.cs:20]: Precisely!

**Comment by sungam3r** [comment]
I just finished.

**Review by sungam3r — APPROVED** [review]

**Comment by sungam3r** [comment]
OK, merge it.

**Pull request merged**